File: 卡码/代码随想录/C++/动态规划/dp/02-背包问题/src/01-01背包/Kama_0046.cpp
//46. 携带研究材料
#include <iostream>
#include <vector>
using namespace std;

int main(){
	int m,n;
	scanf("%d %d",&m,&n);
	int weight[m];
	int value[m];
	for(int i=0;i<m;i++){
		scanf("%d",&weight[i]);
	}
	for(int i=0;i<m;i++){
		scanf("%d",&value[i]);
	}
	vector<vector<int>> dp(m,vector<int>(n+1,0));
	for(int i=weight[0];i<=n;i++) dp[0][i]=value[0];
	for(int i=1;i<m;i++){
		for(int j=1;j<n+1;j++){
			if(j<weight[i]){
				dp[i][j]=dp[i-1][j];
			}else{
				dp[i][j]=max(dp[i-1][j-weight[i]]+value[i],dp[i-1][j]);
			}
		}
	}
	printf("%d\n",dp[m-1][n]);
  system("pause");
  return 0;
}

File: 卡码/代码随想录/1/09-动态规划/Go/02-背包问题/01-01背包/Kama_0046.go
// 46. 携带研究材料
package main

import (
	"fmt"
	"math"
)

func main(){
	var m,n int
	fmt.Scan(&m,&n)
	weight,value:=make([]int,m),make([]int,m)
	for i:=0;i<m;i++ {
		fmt.Scan(&weight[i])
	}
	for i:=0;i<m;i++ {
		fmt.Scan(&value[i])
	}
	dp:=make([][]int,m)
	for i:=0;i<m;i++ {
		dp[i]=make([]int,n+1)
	}
	for i:=weight[0];i<n+1;i++ {
		dp[0][i]=value[0]
	}
	for i:=1;i<m;i++ {
		for j:=1;j<n+1;j++ {
			if j<weight[i] {
				dp[i][j]=dp[i-1][j]
			}else{
				dp[i][j]=int(math.Max(float64(dp[i-1][j]),float64(dp[i-1][j-weight[i]]+value[i])))
			}
		}
	}
	fmt.Println(dp[m-1][n])
}
